File: src/main/java/com/epri/dss/conversion/impl/ISourceObjImpl.java
package com.epri.dss.conversion.impl;

import java.io.PrintStream;

import com.epri.dss.parser.impl.Parser;
import com.epri.dss.shared.impl.CMatrixImpl;
import com.epri.dss.shared.impl.Complex;
import com.epri.dss.shared.impl.ComplexUtil;

import com.epri.dss.common.SolutionObj;
import com.epri.dss.common.impl.DSSClassImpl;
import com.epri.dss.common.impl.DSSGlobals;
import com.epri.dss.common.impl.Utilities;
import com.epri.dss.conversion.ISource;
import com.epri.dss.conversion.ISourceObj;

public class ISourceObjImpl extends PCElementImpl implements ISourceObj {

	private double amps;

	private double angle;

	private double phaseShift;

	private int scanType;

	private int sequenceType;

	protected double srcFrequency;

	public ISourceObjImpl(DSSClassImpl parClass, String sourceName) {
		super(parClass);

		setName(sourceName.toLowerCase());
		objType = parClass.getDSSClassType(); // SOURCE + NON_PCPD_ELEM;  // don't want this in PC Element List

		setNPhases(3);
		nConds = 3;
		setNTerms(1);

		amps     = 0.0;
		angle    = 0.0;
		srcFrequency = baseFrequency;
		phaseShift = 120.0;
		scanType = 1;  // pos sequence
		sequenceType = 1;

		initPropertyValues(0);

		YOrder = nTerms * nConds;
		recalcElementData();
	}

	@Override
	public void recalcElementData() {
		DSSGlobals globals = DSSGlobals.getInstance();

		setSpectrumObj( (com.epri.dss.general.SpectrumObj) globals.getSpectrumClass().find(getSpectrum()) );

		if (getSpectrumObj() == null)
			globals.doSimpleMsg("Spectrum object \"" + getSpectrum() + "\" for device ISource."+getName()+" not found.", 333);

		setInjCurrent( (Complex[]) Utilities.resizeArray(getInjCurrent(), YOrder) );
	}

	@Override
	public void calcYPrim() {

		// build only YPrim_Series
		if (isYprimInvalid()) {
			if (YPrimSeries != null)
				YPrimSeries = null;
			YPrimSeries = new CMatrixImpl(YOrder);
			if (YPrim != null)
				YPrim = null;
			YPrim = new CMatrixImpl(YOrder);
		} else {
			YPrimSeries.clear();
			YPrim.clear();
		}


		/* Yprim = 0 for ideal current source; just leave it zeroed */

		/* Now account for open conductors.
		 * For any conductor that is open, zero out row and column.
		 */
		super.calcYPrim();

		setYPrimInvalid(false);
	}

	private Complex getBaseCurr() {
		double srcHarmonic;
		Complex result = null;

		DSSGlobals globals = DSSGlobals.getInstance();

		try {
			SolutionObj sol = globals.getActiveCircuit().getSolution();

			/* Get first phase current */
			if (sol.isHarmonicModel()) {
				srcHarmonic = sol.getFrequency() / srcFrequency;
				result = getSpectrumObj().getMult(srcHarmonic).multiply(amps);  // base current for this harmonic
				result = Utilities.rotatePhasorDeg(result, srcHarmonic, angle);
			} else {
				if (Math.abs(sol.getFrequency() - srcFrequency) < DSSGlobals.EPSILON2) {
					result = ComplexUtil.polarDeg2Complex(amps, angle);
				} else {
					result = Complex.ZERO;
				}
			}
		} catch (Exception e) {
			globals.doSimpleMsg("Error computing current for ISource."+getName()+". Check specification. Aborting.", 334);
			if (globals.isInRedirect())
				globals.setRedirectAbort(true);
		}

		return result;
	}

	/**
	 * Sum currents directly into solution array.
	 */
	@Override
	public int injCurrents() {
		getInjCurrents(getInjCurrent());

		return super.injCurrents();  // adds into system array
	}

	/**
	 * Total currents into a device.
	 */
	@Override
	public void getCurrents(Complex[] curr) {
		try {
			getInjCurrents(complexBuffer);  // get present value of inj currents
			// add together with YPrim currents
			for (int i = 0; i < YOrder; i++)
				curr[i] = complexBuffer[i].negate();
		} catch (Exception e) {
			DSSGlobals.getInstance().doErrorMsg(("GetCurrents for ISource element: " + getName() + "."), e.getMessage(),
					"Inadequate storage allotted for circuit element.", 335);
		}
	}

	/**
	 * Fill up an array of injection currents.
	 */
	@Override
	public void getInjCurrents(Complex[] curr) {
		SolutionObj sol = DSSGlobals.getInstance().getActiveCircuit().getSolution();

		Complex baseCurr = getBaseCurr();  // this func applies spectrum if needed

		for (int i = 0; i < getNPhases(); i++) {
			curr[i] = baseCurr;
			if (i < getNPhases())
				if (sol.isHarmonicModel()) {
					switch (scanType) {
					case 1:
						baseCurr = Utilities.rotatePhasorDeg(baseCurr, 1.0, -getPhaseShift());  // maintain positive sequence for ISource
						break;
					case 0:
						// do not rotate for zero sequence
						break;
					default:
						baseCurr = Utilities.rotatePhasorDeg(baseCurr, sol.getHarmonic(), -getPhaseShift());  // rotate by frequency
						/* Harmonic 1 will be pos; 2 is neg; 3 is zero, and so on. */
						break;
					}
				} else {
					switch (sequenceType) {
					case -1:
						baseCurr = Utilities.rotatePhasorDeg(baseCurr, 1.0, phaseShift);  // neg seq
						break;
					case 0:
						// do not rotate for zero sequence
						break;
					default:
						baseCurr = Utilities.rotatePhasorDeg(baseCurr, 1.0, -phaseShift);  // maintain pos seq
						break;
					}
				}
		}
	}

	@Override
	public void dumpProperties(PrintStream f, boolean complete) {
		super.dumpProperties(f, complete);

		for (int i = 0; i < getParentClass().getNumProperties(); i++)
			f.println("~ " + getParentClass().getPropertyName()[i] + "=" + getPropertyValue(i));

		if (complete) {
			f.println();
			f.println();
		}
	}

	@Override
	public void initPropertyValues(int arrayOffset) {

		setPropertyValue(0, getBus(1));
		setPropertyValue(1, "0");
		setPropertyValue(2, "0");
		setPropertyValue(3, String.format("%-.6g", srcFrequency));
		setPropertyValue(4, "3");
		setPropertyValue(5, "pos");
		setPropertyValue(6, "pos");

		super.initPropertyValues(ISource.NumPropsThisClass);
	}

	/**
	 * Make a positive sequence model.
	 */
	@Override
	public void makePosSequence() {
		if (getNPhases() > 1) {
			Parser.getInstance().setCmdString("phases=1");
			edit();
		}
		super.makePosSequence();
	}

	public double getSrcFrequency() {
		return srcFrequency;
	}

	public void setSrcFrequency(double frequency) {
		srcFrequency = frequency;
	}

	// FIXME Private members in OpenDSS

	public double getAmps() {
		return amps;
	}

	public void setAmps(double value) {
		amps = value;
	}

	public double getAngle() {
		return angle;
	}

	public void setAngle(double ang) {
		angle = ang;
	}

	public double getPhaseShift() {
		return phaseShift;
	}

	public void setPhaseShift(double shift) {
		phaseShift = shift;
	}

	public int getScanType() {
		return scanType;
	}

	public void setScanType(int type) {
		scanType = type;
	}

	public int getSequenceType() {
		return sequenceType;
	}

	public void setSequenceType(int type) {
		sequenceType = type;
	}

}

File: pydss/conversion/ISourceObjImpl.py
from dss.common.impl.Utilities import (Utilities,)
from dss.conversion.ISourceObj import (ISourceObj,)
from dss.parser.impl.Parser import (Parser,)
from dss.shared.impl.ComplexUtil import (ComplexUtil,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.conversion.impl.PCElementImpl import (PCElementImpl,)
from dss.conversion.ISource import (ISource,)
from dss.shared.impl.CMatrixImpl import (CMatrixImpl,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class ISourceObjImpl(PCElementImpl, ISourceObj):
    amps = None
    angle = None
    phaseShift = None
    scanType = None
    sequenceType = None
    srcFrequency = None

    def __init__(self, parClass, sourceName):
        super(ISourceObjImpl, self)(parClass)
        self.setName(sourceName.toLowerCase())
        self.objType = parClass.getDSSClassType()
        # SOURCE + NON_PCPD_ELEM;  // don't want this in PC Element List
        self.setNPhases(3)
        self.nConds = 3
        self.setNTerms(1)
        self.amps = 0.0
        self.angle = 0.0
        self.srcFrequency = self.baseFrequency
        self.phaseShift = 120.0
        self.scanType = 1
        # pos sequence
        self.sequenceType = 1
        self.initPropertyValues(0)
        self.YOrder = self.nTerms * self.nConds
        self.recalcElementData()

    def recalcElementData(self):
        self.setSpectrumObj(DSSGlobals.spectrumClass.find(self.getSpectrum()))
        if self.getSpectrumObj() is None:
            DSSGlobals.doSimpleMsg('Spectrum object \"' + self.getSpectrum() + '\" for device ISource.' + self.getName() + ' not found.', 333)
        self.setInjCurrent(Utilities.resizeArray(self.getInjCurrent(), self.YOrder))

    def calcYPrim(self):
        # build only YPrim_Series
        if self.isYprimInvalid():
            if self.YPrimSeries is not None:
                self.YPrimSeries = None
            self.YPrimSeries = CMatrixImpl(self.YOrder)
            if self.YPrim is not None:
                self.YPrim = None
            self.YPrim = CMatrixImpl(self.YOrder)
        else:
            self.YPrimSeries.clear()
            self.YPrim.clear()
        # Yprim = 0 for ideal current source; just leave it zeroed
        # Now account for open conductors.
        # For any conductor that is open, zero out row and column.

        super(ISourceObjImpl, self).calcYPrim()
        self.setYPrimInvalid(False)

    def getBaseCurr(self):
        result = None
        try:
            sol = DSSGlobals.activeCircuit.getSolution()
            # Get first phase current
            if sol.isHarmonicModel():
                srcHarmonic = sol.getFrequency() / self.srcFrequency
                result = self.getSpectrumObj().getMult(srcHarmonic).multiply(self.amps)
                # base current for this harmonic
                result = Utilities.rotatePhasorDeg(result, srcHarmonic, self.angle)
            elif (
                self.Math.abs(sol.getFrequency() - self.srcFrequency) < DSSGlobals.EPSILON2
            ):
                result = ComplexUtil.polarDeg2Complex(self.amps, self.angle)
            else:
                result = Complex.ZERO
        except Exception, e:
            DSSGlobals.doSimpleMsg('Error computing current for ISource.' + self.getName() + '. Check specification. Aborting.', 334)
            if DSSGlobals.inRedirect:
                DSSGlobals.redirectAbort = True
        return result

    def injCurrents(self):
        """Sum currents directly into solution array."""
        self.getInjCurrents(self.getInjCurrent())
        return super(ISourceObjImpl, self).injCurrents()
        # adds into system array

    def getCurrents(self, curr):
        """Total currents into a device."""
        try:
            self.getInjCurrents(self.complexBuffer)
            # get present value of inj currents
            # add together with YPrim currents
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.YOrder):
                    break
                curr[i] = self.complexBuffer[i].negate()
        except Exception, e:
            DSSGlobals.doErrorMsg('GetCurrents for ISource element: ' + self.getName() + '.', e.getMessage(), 'Inadequate storage allotted for circuit element.', 335)

    def getInjCurrents(self, curr):
        """Fill up an array of injection currents."""
        sol = DSSGlobals.activeCircuit.getSolution()
        baseCurr = self.getBaseCurr()
        # this func applies spectrum if needed
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getNPhases()):
                break
            curr[i] = baseCurr
            if i < self.getNPhases():
                if sol.isHarmonicModel():
                    _1 = self.scanType
                    _2 = False
                    while True:
                        if _1 == 1:
                            _2 = True
                            baseCurr = Utilities.rotatePhasorDeg(baseCurr, 1.0, -self.getPhaseShift())
                            # maintain positive sequence for ISource
                            break
                        if (_2 is True) or (_1 == 0):
                            _2 = True
                            break
                        if True:
                            _2 = True
                            baseCurr = Utilities.rotatePhasorDeg(baseCurr, sol.getHarmonic(), -self.getPhaseShift())
                            # rotate by frequency
                            # Harmonic 1 will be pos; 2 is neg; 3 is zero, and so on.
                            break
                        break
                else:
                    _3 = self.sequenceType
                    _4 = False
                    while True:
                        if _3 == -1:
                            _4 = True
                            baseCurr = Utilities.rotatePhasorDeg(baseCurr, 1.0, self.phaseShift)
                            # neg seq
                            break
                        if (_4 is True) or (_3 == 0):
                            _4 = True
                            break
                        if True:
                            _4 = True
                            baseCurr = Utilities.rotatePhasorDeg(baseCurr, 1.0, -self.phaseShift)
                            # maintain pos seq
                            break
                        break

    def dumpProperties(self, f, complete):
        super(ISourceObjImpl, self).dumpProperties(f, complete)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getParentClass().getNumProperties()):
                break
            print '~ ' + self.getParentClass().getPropertyName()[i] + '=' + self.getPropertyValue(i)
        if complete:
            print 
            print 

    def initPropertyValues(self, arrayOffset):
        self.setPropertyValue(0, self.getBus(0))
        self.setPropertyValue(1, '0')
        self.setPropertyValue(2, '0')
        self.setPropertyValue(3, String.format.format('%-.6g', self.srcFrequency))
        self.setPropertyValue(4, '3')
        self.setPropertyValue(5, 'pos')
        self.setPropertyValue(6, 'pos')
        super(ISourceObjImpl, self).initPropertyValues(ISource.NumPropsThisClass)

    def makePosSequence(self):
        """Make a positive sequence model."""
        if self.getNPhases() > 1:
            Parser.getInstance().setCmdString('phases=1')
            self.edit()
        super(ISourceObjImpl, self).makePosSequence()

    def getSrcFrequency(self):
        return self.srcFrequency

    def setSrcFrequency(self, frequency):
        # FIXME Private members in OpenDSS
        self.srcFrequency = frequency

    def getAmps(self):
        return self.amps

    def setAmps(self, value):
        self.amps = value

    def getAngle(self):
        return self.angle

    def setAngle(self, ang):
        self.angle = ang

    def getPhaseShift(self):
        return self.phaseShift

    def setPhaseShift(self, shift):
        self.phaseShift = shift

    def getScanType(self):
        return self.scanType

    def setScanType(self, type):
        self.scanType = type

    def getSequenceType(self):
        return self.sequenceType

    def setSequenceType(self, type):
        self.sequenceType = type
